### PR TITLE
Add pure sine tone generator

### LINKED ISSUE #736: Signal generators: tone and chirp
#### Description
It's often useful to be able to synthesize a pure sine tone, or either a linear or exponential sine sweep.  It's not hard to implement these by hand, but it would be nice to have them as primitives in the library, much like we currently have ``librosa.click``.

For signatures, I'm thinking:

```python
tone(frequency, sr, length=None, duration=None)
```
- The first two are obvious (Hz)
- `length` would be in samples, as in `istft`, `click`, or the effects module.
- `duration` would allow the user to specify the duration in seconds rather than samples, though `length` would take priority.  At least one of `lenth` or `duration` must be provided.

```python
chrip(fmin, fmax, sr, length=None, duration=None, linear=False)
```
- The first 5 are as above
- `linear` is a boolean that toggles between linear and exponential sweep.

### PR DESCRIPTION
**Pull request by ZiyaoWei**
#### Reference Issue
Partially fixes #736
#### What does this implement/fix? Explain your changes.
This implements the pure sine tone signal generator as specified in #736.
#### Any other comments?
None

---
This change is [Reviewable](https://reviewable.io/reviews/librosa/librosa/753)

### CODE REVIEW

**Comment by lostanlen** [comment]
Hello @ZiyaoWei, thank you for doing this. It would be good if you could add a `phase` keyword argument, defaulting to zero.

**Comment by lostanlen** [comment]
Also, I would recommend that the sine wave uses `np.cos` rather than `np.sin`. That way, the default case of `phase=0` will make the wave symmetric.

**Comment by ZiyaoWei** [comment]
@lostanlen - thanks for the suggestion - I'll definitely add the `phi` parameter (it should be using degrees and not radians I assume?)! For my edification - why is `cos` + 90 degrees preferred to `-sin`?

**Comment by lostanlen** [comment]
Yes, `phi` is a better name for the keyword that what I proposed earlier (`phase`), given that it matches `scipy.signal.chirp`.

I would prefer `cos` over `sin` because `cos(omega t)` is the real part of `exp(i omega t)`. That way, `phi` is equal the Hilbert analytic phase of the tone.

See Eq 4.47 of Mallat's Wavelet Tour: http://links.uwaterloo.ca/amath391w13docs/Mallat3.pdf
and this paper of Flandrin http://perso.ens-lyon.fr/patrick.flandrin/SPIE01_PF.pdf

**Comment by ZiyaoWei** [comment]
@lostanlen - thanks for the explanation! That's quite a bit more technical than I thought :-)

**Comment by ZiyaoWei** [comment]
(Updated - I amended the commit so not sure if GitHub will show it as updated or not)

**Comment by bmcfee** [comment]
This looks great, thanks for doing it so quickly!  And thanks @lostanlen for the input.

**Comment by bmcfee** [comment]
@ZiyaoWei Hey, @lostanlen and I discussed this offline, and think it would be better to use radians rather than degrees for `phi`, both here and in #750.  This disagrees with `scipy.signal`'s API, but elsewhere in librosa, phase is always in radians (or complex exponential) as produced by `np.angle`.

Would you mind making that change in both places?  To keep the definition clean, I'd suggest using `phi=None` as a default, and converting `None` to `-np.pi * 0.5` at runtime; this would keep us from having a long string of numbers in the function signature.  Also, if the convention here is `cos(x + phi)`, then I'd prefer a default `phi = - np.pi * 0.5` so that the signal starts with positive slope (equiv. to sine) rather than negative.  This is still consistent with @lostanlen's argument for `cos(x + phi)`, but still gives us the expected sine-like behavior by default.

Thanks!

**Comment by ZiyaoWei** [comment]
@bmcfee Sure, will make that change today - I will also add the return value shape in the documentation too a la #750 , thanks for reviewing!

File: librosa/core/audio.py [hunk 1]
```diff
@@ -712,3 +712,66 @@ def clicks(times=None, frames=None, sr=22050, hop_length=512,
             click_signal[start:end] += click
 
     return click_signal
+
+
+def tone(frequency, sr=22050, length=None, duration=None, phi=90):
```
Comment (bmcfee) [inline librosa/core/audio.py]: (Redundant note here, just for maintaining GH status)

1. phi should be in radians, not degrees
2. default should be `None`
3. `phi=None` should map to `-np.pi/2` (not `+np.pi/2`)

**Comment by ZiyaoWei** [comment]
Didn't quite keep the promise - only got around to it now :-/ Will update `chirp` shortly too.

**Comment by bmcfee** [comment]
Thanks!  I think this one's good to merge.

I suspect this will cause a merge conflict on #750, so maybe wait a minute and rebase master after I merge this?

**Comment by lostanlen** [comment]
Congratulations on your merged PR @ZiyaoWei! I'm sorry that I gave you the runaround with that `phi` kwarg. This looks great now.
We'll make sure to mention the addition of `librosa.tone` as a new feature to v0.6.2 in the release notes, to give it some publicity.

**Pull request merged**